### DISCUSSION TITLE
test client

### DISCUSSION THREAD

**Discussion by lawndoc** [miguelgrinberg/flask-sock#88] [Q&A]
Does this Flask extension work with Flask's built in test client? Or does that need to be implemented in this project?

**Comment by lawndoc** [comment]
Based on my (failed) testing, it looks like I'll need a `Sock.test_client(app)` method for what I'm trying to do. Unless I'm doing something wrong.

If you are busy with other things, I can try to help implement it. If you know of examples in other repos that I can reference that would be a big help.

Edit: maybe something like your [Flask-SocketIO test client](https://github.com/miguelgrinberg/Flask-SocketIO/blob/main/src/flask_socketio/test_client.py)?

**Comment by lawndoc** [comment]
I did look into using Flask-SocketIO to get the test client functionality, but [my use case](https://github.com/lawndoc/commander) requires a long-lived, fully duplexed connection so I can perform a server push when something happens on the backend (not socket event-based).

This seemed like the easier way to go because I'm much more familiar with socket programming than trying to get that to work with socket.io. I also want to keep the particular component I'm referring to as lean as possible because it will have the greatest impact on server performance.

**Comment by miguelgrinberg** [comment]
There is no support for the test client at this time, sorry.

You can implement your own testing set up, for example by calling your websocket endpoint with a mocked `ws` object that allows you to carry out your tests.

**Comment by lawndoc** [comment]
Gonna leave a helpful message here in case anyone is trying to do the same thing as me. It might spare them the time it took me to figure all this out.

You're going to end up needing to bypass the `@sock.route()` decorator on your function to test it, or you'll get some Flask error about the environment not being valid when you call the function. To call the function directly and bypass the decorator, you just need to *import the function you're testing into your test script* and *call its `__wrapped__` attribute*.

Since we don't have the decorator creating and passing in a `simple_websocket.Server` object, you also need to *implement your own mock websocket server object* that can perform the `send` and `recv` calls in your function. Miguel mentioned this in his above comment.

Here's an example flask-sock route and a unit test for it:

*routes.py*
```python
@sock.route("/api/ws")
def websockRoute(ws):
    while True:
        msg = ws.receive()
        if msg == "foo":
            ws.send("bar")
        try:
            return ws.mockServer
        except AttributeError:
            pass
```

*test_routes.py*
```python
from routes import websockRoute

class MockServer:
    def __init__(self):
        self.lastMessage = None
        self.mockServer = True
    def send(self, msg):
        self.lastMessage = msg
    def receive(self):
        return "foo"
    def close(self, *args, **kwargs):
        return

def testWebsockRoute():
    websockRoute.__wrapped__(MockServer())
```

I found it helpful to *keep track of the last message* that was sent by the route function for my use case. You'll need to *modify the sends and receives* to create valid (or intentionally invalid) data for the function you're testing.

I also added the `mockServer` attribute to our mock `ws` object so we can stop the while loop during testing, but continue to iterate when actually running the Flask app.

**Comment by miguelgrinberg** [comment]
Thanks for posting your solution for others to see!

I do not recommend that you access the `__wrapped__` element, though. You are using private parts of the Flask and Flask-Sock implementations that can change at any time.

The preferred solution is to call the function directly, as Flask does when it dispatches a request. I think all you need to do to avoid errors is to install a test request context before you call the function.

**Comment by snarfed** [comment]
As another option, here's a simple fake I'm currently using, at the simple_websocket level:

```py
class FakeWebsocketClient:
    """Fake of :class:`simple_websocket.Client`."""
    url = None
    sent = []
    to_receive = []

    def __init__(self, url):
        FakeWebsocketClient.url = url

    def send(self, msg):
        self.sent.append(json.loads(msg))

    def receive(self):
        if not self.to_receive:
            raise simple_websocket.ConnectionClosed(message='foo')


class ClientTest(TestCase):
    def setUp(self):
        simple_websocket.Client = FakeWebsocketClient
        FakeWebsocketClient.url = None
        FakeWebsocketClient.sent = []
        FakeWebsocketClient.to_receive = []

    def test_something(self):
        FakeWebsocketClient.to_receive = [{'num': 3}, {'num': 4}]

        # run my code

        self.assertEqual('http://expected/...', FakeWebsocketClient.url)
        # ...
```
